Add scoring functions and new gene info

elenaara opened the pull request:
I added a new script with scoring functions, should be merged with the current predict_functions script, with the new main predicting funcion replacing the old one.
A new plotting function that shows the additional scores should also be added (the idea was to have a plot that shows only the scores).

elenaara committed:
Add scoring functions and new gene info

mattssca commented:
Thanks for adding this. I have had a first look at it. Before I am adding a more comprehensive review, including testing, one thing that comes to mind is the following. I see many functions starting with the prefix `score_` and they seem to have reoccurring parameters such as `Data`, `logTransformed`, `gene_id` and `method`. 

Would it be possible and make sense to generalize these functions? I.e would it be make sense to boil these down to one funciton with an added parameter. For example a new parameter called `variable` that holds information about what score (progression, prostate, grade. immune, etc.)?

In addition, would you like me to add the additional datasets to the R package with `use_this::` so they end up in the data folder, or are you still working on these datasets?

Lastly, if the idea is that this new script will replace the existing `predict_functions.R` you can go ahead and delete this from your branch.

Thanks!

elenaara committed:
Update scoring functions, signatures and gene info datasets

elenaara committed:
Add visualize_scores

elenaara commented:
Thank you for all your suggestions!

I merged the scores that are calculated in the same way into two functions, ratio_score and single_score for the ones that are calculated using two sets of genes (up and down) or one, respectively. I moved these functions to the predict_functions.R script, thinking of moving all the finished functions there.

For the addional data that is needed, I am uploading the signatures and the gene info files. In the previous code each score had the needed genes in a separate file, but the new signatures file includes everything. If there is anything else that you see it's needed you can upload it. The grade classifiers come to mind, but I don't know if there is anything else.

I am also uploading the visualize_scores script that includes some code to generate the score plots. It is a bit messy, not even a function, so I hope you can make some sense of it. Of course ask me if there is anything unclear.

elenaara committed:
Update predicting function and add to script

elenaara committed:
Update plotting function

elenaara committed:
Update links to gene info

mattssca commented:
# PR Review

## visualize_scores.R causes error with `devtools::load_all()`
You can delete this script since I have already downloaded a copy of it to my local computer.

## Predict_LundTax2023(Lund2017) throws an error;

```
> results <- predict_LundTax2023(Lund2017)
Error in predict_LundTax2023(Lund2017) : object 'verbose' not found
```

This error is returned because in `predict_RF` the verbose parameter is set to TRUE (default) and in `lund_scores` the same parameter has FALSE as the default value. Inside `predict_LundTax2023` the verbose option for the internal call of `predict_RF` is hardcoded to TRUE. The best fix for this is to add the `verbose` parameter to `predict_LundTax2023` and assign the input of this parameter to the internal call of `predict_RF` in that function. I can confirm that this works.

## Referring to the Signatures object as updated_signatures.
There are many referrals to the `signatures` object as `updated_signatures`. I think this is a remnant from when you used an updated version of the gene signatures. All lines in the code should be updated to refer to the bundled object (i.e. signatures). As a side note, do we need to update the signatures object? If so, I can create an issue of this for tracking purposes.

As a quick workaround for my testing purposes, I fixed this error by simply assigning the signatures object a new name as such;

```
updated_signatures = LundTax2023Classifier::signatures
```

## The hardcoded paths to needed datasets need to be updated.
There are a few instances where the package is trying to source files locally stored on the desktop computer in the lab. These are the objects I added to my PR yesterday. The code should load these objects from the package and local file paths removed.

```
> results <- predict_LundTax2023(Lund2017)
[1] "Calculating Proliferation"
Proliferation score: 6/138 genes are missing from the data.
[1] "NUSAP1"   "PRC1"     "FAM72D"   "FAM72A"   "RNASEH2A" "POLA2"   
Error in readChar(con, 5L, useBytes = TRUE) : cannot open the connection
In addition: Warning message:
In readChar(con, 5L, useBytes = TRUE) :
  cannot open compressed file 'D:/UROSCANSEQ_2024/Analysis/02.New_data/GradeClassifier/RF/hyperparameterCV/CLASSIFIER_RF_Grade.RData', probable reason 'No such file or directory'
```

## More hardcoded paths.
Similarly, the package also tries to read from the following path; `C:/Users/earam/LBCG/gene_info_lund.rda`. This should also be updated to use the bundled object and if the object needs to be updated, this should also be done.

```
> results <- predict_LundTax2023(Lund2017)
[1] "Calculating Proliferation"
Proliferation score: 6/138 genes are missing from the data.
[1] "NUSAP1"   "PRC1"     "FAM72D"   "FAM72A"   "RNASEH2A" "POLA2"   
[1] "Calculating G3 score"
Error in readChar(con, 5L, useBytes = TRUE) : cannot open the connection
In addition: Warning message:
In readChar(con, 5L, useBytes = TRUE) :
  cannot open compressed file 'C:/Users/earam/LBCG/gene_info_lund.rda', probable reason 'No such file or directory'
```

## The test code produces an error.
After working around the above-mentioned errors, `results <- predict_LundTax2023(Lund2017)` returns an error. This is because some genes are missing. It is possible to run this with `impute = TRUE` and the code runs. But I think this error is related to the updated gene list in my previous point. Or if not, the example code should be updated to have the `impute = TRUE` argument.

```
> results <- predict_LundTax2023(Lund2017)
[1] "Calculating Proliferation"
Proliferation score: 6/138 genes are missing from the data.
[1] "NUSAP1"   "PRC1"     "FAM72D"   "FAM72A"   "RNASEH2A" "POLA2"   
[1] "Calculating G3 score"
Loading required package: multiclassPairs
These genes are not found in the data:
ENSG00000109654 ENSG00000290292 ENSG00000115561 ENSG00000182768 ENSG00000139372 ENSG00000102172 ENSG00000292348 ENSG00000188610 ENSG00000125885 ENSG00000196550 ENSG00000013364 ENSG00000120963 ENSG00000173207 ENSG00000137804 ENSG00000133794 ENSG00000280789
Gene names should as rownames and sample names as columns!
Check the genes in classifier object to see all the needed genes.
Check if '-' or ',' symbols in the gene names in your data. You may need to change it to '_' or '.'
Error in predict_RF(classifier = grade_predictor, Data = D, ...) : 
  All genes should be in the data with no NA values! Or you can turn impute argument to TRUE to impute missed genes to the closest class for each sample!
```

## Wrong name for the first column in the returned scores data frame.
In the returned list, looking at the scores data frame, the first column has the wrong name. It is returned as `Score`;

```
results <- predict_LundTax2023(Lund2017, impute = TRUE)
colnames(results$scores)[1]
[1] "Score"
```

The correct name for this column is `ProliferationScore`, it's important since the column is accessed elsewhere in the code and the correct column name is needed.

## Potentially inaccurate values in the scores data frame.
I think the values in the returned scores data frame needs to be carefully looked into to ensure the output is what is expected. For example, 

```
> table(results$scores$PossibleProstate)

YES 
301 
```

```
> head(results$scores$Immune141_UP)
[1] NaN NaN NaN NaN NaN NaN
```

```
> head(results$scores$Stromal141_UP)
[1] NaN NaN NaN NaN NaN NaN
```

For your convenience, here is the head of the scores data frame. The best is probably to look into these values and see if they biologically make sense to you.

```
> head(results$scores)
                Score MolecularGradeWHO1999 MolecularGradeWHO1999_score MolecularGradeWHO2016 MolecularGradeWHO2016_score
p1404_1.CEL 0.7310869                    G3                   0.8074286                    HG                   0.7736667
2.CEL       1.0419366                    G3                   0.8580000                    HG                   0.7456667
3.CEL       0.4846604                    G3                   0.7255000                    HG                   0.6576667
4.CEL       0.6596973                    G3                   0.8650000                    HG                   0.8116667
5.CEL       0.8079496                    G3                   0.7266667                    HG                   0.7781667
6.CEL       0.9343730                    G3                   0.8390000                    HG                   0.7670000
            ProgressionScore ProgressionRisk ProstateScore PossibleProstate Immune141_UP  B-cells  T-cells T-cells CD8+ NK-cells
p1404_1.CEL        0.7675475              HR      3.762919              YES          NaN 3.211749 3.569395     1.966563 3.162512
2.CEL              0.9923554              HR      3.984649              YES          NaN 3.443050 4.281772     2.077998 3.451555
3.CEL              0.5824121              HR      3.999140              YES          NaN 4.439361 5.056852     2.795713 3.455866
4.CEL              0.7804901              HR      4.027033              YES          NaN 3.329162 3.695808     2.284374 3.365593
5.CEL              0.6303658              HR      3.621453              YES          NaN 3.116570 3.727694     2.270413 3.290665
6.CEL              0.8631717              HR      3.714340              YES          NaN 3.195027 3.802479     1.991293 3.197117
            Cytotoxicity Score Neutrophils Monocytic lineage Macrophages M2 macrophage Myeloid Dendritic Cells Stromal141_UP
p1404_1.CEL           3.530327    3.405826          4.281524    4.844223      3.806195                2.997839           NaN
2.CEL                 4.159873    4.063120          6.034082    6.992902      5.932902                3.447133           NaN
3.CEL                 4.150582    3.764249          5.015871    5.509321      4.348992                3.901183           NaN
4.CEL                 3.743617    3.598783          4.513464    5.178399      3.883522                3.099180           NaN
5.CEL                 3.655186    3.651167          4.267284    4.930646      3.435998                3.286593           NaN
6.CEL                 3.673182    3.520885          4.291904    4.932714      3.659754                3.028265           NaN
            Endothelial cells Fibroblasts Smooth muscle B-cells Proportion T-cells Proportion T-cells CD8+ Proportion
p1404_1.CEL          3.781402    6.999977      6.360039         0.06186246         0.06875119              0.03787857
2.CEL                4.208606    7.738854      6.389830         0.05533522         0.06881479              0.03339669
3.CEL                4.241815    7.489493      7.581455         0.07189161         0.08189133              0.04527415
4.CEL                3.923774    6.361952      6.167985         0.06264227         0.06954117              0.04298331
5.CEL                3.857191    6.142944      5.650759         0.06077186         0.07268853              0.04427215
6.CEL                4.329897    6.518604      5.913825         0.06137798         0.07304742              0.03825367
            NK-cells Proportion Cytotoxicity Score Proportion Neutrophils Proportion Monocytic lineage Proportion
p1404_1.CEL          0.06091409                    0.06799869             0.06560065                   0.08246773
2.CEL                0.05547191                    0.06685569             0.06530071                   0.09697717
3.CEL                0.05596475                    0.06721508             0.06095876                   0.08122769
4.CEL                0.06332778                    0.07044076             0.06771552                   0.08492636
5.CEL                0.06416664                    0.07127466             0.07119628                   0.08321033
6.CEL                0.06141811                    0.07056355             0.06763785                   0.08244950
            Macrophages Proportion M2 macrophage Proportion Myeloid Dendritic Cells Proportion Endothelial cells Proportion
p1404_1.CEL             0.09330604               0.07331228                         0.05774229                   0.07283472
2.CEL                   0.11238690               0.09535104                         0.05540084                   0.06763891
3.CEL                   0.08921869               0.07042817                         0.06317627                   0.06869252
4.CEL                   0.09743794               0.07307324                         0.05831487                   0.07383063
5.CEL                   0.09614560               0.06700058                         0.06408724                   0.07521367
6.CEL                   0.09475976               0.07030559                         0.05817439                   0.08317936
            Fibroblasts Proportion Smooth muscle Proportion
p1404_1.CEL              0.1348287                0.1225026
2.CEL                    0.1243755                0.1026946
3.CEL                    0.1212859                0.1227751
4.CEL                    0.1197079                0.1160582
5.CEL                    0.1197849                0.1101875
6.CEL                    0.1252254                0.1136074
```

mattssca commented:
I've also tried the implemented plotting function and it returns an error related to incorrect  number of dimension, this is the output:

```
> results <- predict_LundTax2023(Lund2017, 
+                                include_data = TRUE, 
+                                impute = TRUE, 
+                                verbose = TRUE)

[1] "Calculating Proliferation"
Proliferation score: 6/138 genes are missing from the data.
[1] "NUSAP1"   "PRC1"     "FAM72D"   "FAM72A"   "RNASEH2A" "POLA2"   
[1] "Calculating G3 score"
Loading required package: multiclassPairs
These genes are not found in the data:
ENSG00000109654 ENSG00000290292 ENSG00000115561 ENSG00000182768 ENSG00000139372 ENSG00000102172 ENSG00000292348 ENSG00000188610 ENSG00000125885 ENSG00000196550 ENSG00000013364 ENSG00000120963 ENSG00000173207 ENSG00000137804 ENSG00000133794 ENSG00000280789
Gene names should as rownames and sample names as columns!
Check the genes in classifier object to see all the needed genes.
Check if '-' or ',' symbols in the gene names in your data. You may need to change it to '_' or '.'
Missed genes will be imputed to the closest class for each sample!
These genes have NAs:
ENSG00000109654 ENSG00000290292 ENSG00000115561 ENSG00000182768 ENSG00000139372 ENSG00000102172 ENSG00000292348 ENSG00000188610 ENSG00000125885 ENSG00000196550 ENSG00000013364 ENSG00000120963 ENSG00000173207 ENSG00000137804 ENSG00000133794 ENSG00000280789
These genes will be imputed to the closest class for each sample with NAs
[1] "Calculating HG score"
Loading required package: multiclassPairs
These genes are not found in the data:
ENSG00000224531 ENSG00000104889 ENSG00000108106 ENSG00000172236 ENSG00000147155 ENSG00000198300 ENSG00000125885 ENSG00000149300 ENSG00000284479 ENSG00000069764 ENSG00000120669 ENSG00000171489 ENSG00000113522 ENSG00000228727 ENSG00000187838 ENSG00000269955
Gene names should as rownames and sample names as columns!
Check the genes in classifier object to see all the needed genes.
Check if '-' or ',' symbols in the gene names in your data. You may need to change it to '_' or '.'
Missed genes will be imputed to the closest class for each sample!
These genes have NAs:
ENSG00000224531 ENSG00000104889 ENSG00000108106 ENSG00000172236 ENSG00000147155 ENSG00000198300 ENSG00000125885 ENSG00000149300 ENSG00000284479 ENSG00000069764 ENSG00000120669 ENSG00000171489 ENSG00000113522 ENSG00000228727 ENSG00000187838 ENSG00000269955
These genes will be imputed to the closest class for each sample with NAs
[1] "Calculating Progression score"
Progression score: 28/200 genes are missing from the data.
 [1] "CCDC150"       "TATDN2"        "FAM72A"        "FAM72B"        "ZNF706"        "ESPL1"         "LOXL2"         "PCDHB3"       
 [9] "ZNF670-ZNF695" "UBE2D3"        "TSPYL1"        "EEF1B2"        "LY75"          "RPL26"         "EEF1A1"        "ZNF816"       
[17] "TMBIM4"        "RPL27A"        "PFDN5"         "RPS27"         "RPS3A"         "EIF4B"         "TPT1"          "SLC35A1"      
[25] "RPL13A"        "LDHB"          "EEF1G"         "BORCS7"       
[1] "Calculating Prostate score"
Prostate score: 3/17 genes are missing from the data.
[1] "POTEI"  "NKX3-1" "PDE11A"
[1] "Calculating Immune scores"
Immune scores: 3/71 genes are missing from the data.
[1] "KIR2DL3" "KLRK1"   "FCGR3B" 
[1] "Merging scores"

> plot_signatures(results)
Error in signatures[which(signatures$Signature == "early_cell_cycle."),  : 
  incorrect number of dimensions
```

mattssca commented:
One more thing, could you update my role permissions in this group so I can access the settings of this repo? You can find out how to do this [here](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-teams-and-people-with-access-to-your-repository#changing-permissions-for-a-team-or-person)

mattssca commented on R/predict_functions.R:
```diff
@@ -1,806 +1,1339 @@
+                                logTransform = FALSE,
+                                adjust = TRUE, # adjust scores by stable genes and adjustment factor
+                                adj_factor = 5.1431,# adjustment factor
+                                ...)
```
Add `verbose = TRUE` here

mattssca reviewed (changes requested):
Thanks for adding this! I just have a few minor comments related to the updates introduced in this PR.

## visualize_scores.R causes error with `devtools::load_all()`
You can delete this script since I have already downloaded a copy of it to my local computer.

## Predict_LundTax2023(Lund2017) throws an error;

```
> results <- predict_LundTax2023(Lund2017)
Error in predict_LundTax2023(Lund2017) : object 'verbose' not found
```

This error is returned because in `predict_RF` the verbose parameter is set to TRUE (default) and in `lund_scores` the same parameter has FALSE as the default value. Inside `predict_LundTax2023` the verbose option for the internal call of `predict_RF` is hardcoded to TRUE. The best fix for this is to add the `verbose` parameter to `predict_LundTax2023` and assign the input of this parameter to the internal call of `predict_RF` in that function. I can confirm that this works.

## Referring to the Signatures object as updated_signatures.
There are many referrals to the `signatures` object as `updated_signatures`. I think this is a remnant from when you used an updated version of the gene signatures. All lines in the code should be updated to refer to the bundled object (i.e. signatures). As a side note, do we need to update the signatures object? If so, I can create an issue of this for tracking purposes.

As a quick workaround for my testing purposes, I fixed this error by simply assigning the signatures object a new name as such;

```
updated_signatures = LundTax2023Classifier::signatures
```

## The hardcoded paths to needed datasets need to be updated.
There are a few instances where the package is trying to source files locally stored on the desktop computer in the lab. These are the objects I added to my PR yesterday. The code should load these objects from the package and local file paths removed.

```
> results <- predict_LundTax2023(Lund2017)
[1] "Calculating Proliferation"
Proliferation score: 6/138 genes are missing from the data.
[1] "NUSAP1"   "PRC1"     "FAM72D"   "FAM72A"   "RNASEH2A" "POLA2"   
Error in readChar(con, 5L, useBytes = TRUE) : cannot open the connection
In addition: Warning message:
In readChar(con, 5L, useBytes = TRUE) :
  cannot open compressed file 'D:/UROSCANSEQ_2024/Analysis/02.New_data/GradeClassifier/RF/hyperparameterCV/CLASSIFIER_RF_Grade.RData', probable reason 'No such file or directory'
```

## More hardcoded paths.
Similarly, the package also tries to read from the following path; `C:/Users/earam/LBCG/gene_info_lund.rda`. This should also be updated to use the bundled object and if the object needs to be updated, this should also be done.

```
> results <- predict_LundTax2023(Lund2017)
[1] "Calculating Proliferation"
Proliferation score: 6/138 genes are missing from the data.
[1] "NUSAP1"   "PRC1"     "FAM72D"   "FAM72A"   "RNASEH2A" "POLA2"   
[1] "Calculating G3 score"
Error in readChar(con, 5L, useBytes = TRUE) : cannot open the connection
In addition: Warning message:
In readChar(con, 5L, useBytes = TRUE) :
  cannot open compressed file 'C:/Users/earam/LBCG/gene_info_lund.rda', probable reason 'No such file or directory'
```

## The test code produces an error.
After working around the above-mentioned errors, `results <- predict_LundTax2023(Lund2017)` returns an error. This is because some genes are missing. It is possible to run this with `impute = TRUE` and the code runs. But I think this error is related to the updated gene list in my previous point. Or if not, the example code should be updated to have the `impute = TRUE` argument.

```
> results <- predict_LundTax2023(Lund2017)
[1] "Calculating Proliferation"
Proliferation score: 6/138 genes are missing from the data.
[1] "NUSAP1"   "PRC1"     "FAM72D"   "FAM72A"   "RNASEH2A" "POLA2"   
[1] "Calculating G3 score"
Loading required package: multiclassPairs
These genes are not found in the data:
ENSG00000109654 ENSG00000290292 ENSG00000115561 ENSG00000182768 ENSG00000139372 ENSG00000102172 ENSG00000292348 ENSG00000188610 ENSG00000125885 ENSG00000196550 ENSG00000013364 ENSG00000120963 ENSG00000173207 ENSG00000137804 ENSG00000133794 ENSG00000280789
Gene names should as rownames and sample names as columns!
Check the genes in classifier object to see all the needed genes.
Check if '-' or ',' symbols in the gene names in your data. You may need to change it to '_' or '.'
Error in predict_RF(classifier = grade_predictor, Data = D, ...) : 
  All genes should be in the data with no NA values! Or you can turn impute argument to TRUE to impute missed genes to the closest class for each sample!
```

## Wrong name for the first column in the returned scores data frame.
In the returned list, looking at the scores data frame, the first column has the wrong name. It is returned as `Score`;

```
results <- predict_LundTax2023(Lund2017, impute = TRUE)
colnames(results$scores)[1]
[1] "Score"
```

The correct name for this column is `ProliferationScore`, it's important since the column is accessed elsewhere in the code and the correct column name is needed.

## Potentially inaccurate values in the scores data frame.
I think the values in the returned scores data frame needs to be carefully looked into to ensure the output is what is expected. For example, 

```
> table(results$scores$PossibleProstate)

YES 
301 
```

```
> head(results$scores$Immune141_UP)
[1] NaN NaN NaN NaN NaN NaN
```

```
> head(results$scores$Stromal141_UP)
[1] NaN NaN NaN NaN NaN NaN
```

For your convenience, here is the head of the scores data frame. The best is probably to look into these values and see if they biologically make sense to you.

```
> head(results$scores)
                Score MolecularGradeWHO1999 MolecularGradeWHO1999_score MolecularGradeWHO2016 MolecularGradeWHO2016_score
p1404_1.CEL 0.7310869                    G3                   0.8074286                    HG                   0.7736667
2.CEL       1.0419366                    G3                   0.8580000                    HG                   0.7456667
3.CEL       0.4846604                    G3                   0.7255000                    HG                   0.6576667
4.CEL       0.6596973                    G3                   0.8650000                    HG                   0.8116667
5.CEL       0.8079496                    G3                   0.7266667                    HG                   0.7781667
6.CEL       0.9343730                    G3                   0.8390000                    HG                   0.7670000
            ProgressionScore ProgressionRisk ProstateScore PossibleProstate Immune141_UP  B-cells  T-cells T-cells CD8+ NK-cells
p1404_1.CEL        0.7675475              HR      3.762919              YES          NaN 3.211749 3.569395     1.966563 3.162512
2.CEL              0.9923554              HR      3.984649              YES          NaN 3.443050 4.281772     2.077998 3.451555
3.CEL              0.5824121              HR      3.999140              YES          NaN 4.439361 5.056852     2.795713 3.455866
4.CEL              0.7804901              HR      4.027033              YES          NaN 3.329162 3.695808     2.284374 3.365593
5.CEL              0.6303658              HR      3.621453              YES          NaN 3.116570 3.727694     2.270413 3.290665
6.CEL              0.8631717              HR      3.714340              YES          NaN 3.195027 3.802479     1.991293 3.197117
            Cytotoxicity Score Neutrophils Monocytic lineage Macrophages M2 macrophage Myeloid Dendritic Cells Stromal141_UP
p1404_1.CEL           3.530327    3.405826          4.281524    4.844223      3.806195                2.997839           NaN
2.CEL                 4.159873    4.063120          6.034082    6.992902      5.932902                3.447133           NaN
3.CEL                 4.150582    3.764249          5.015871    5.509321      4.348992                3.901183           NaN
4.CEL                 3.743617    3.598783          4.513464    5.178399      3.883522                3.099180           NaN
5.CEL                 3.655186    3.651167          4.267284    4.930646      3.435998                3.286593           NaN
6.CEL                 3.673182    3.520885          4.291904    4.932714      3.659754                3.028265           NaN
            Endothelial cells Fibroblasts Smooth muscle B-cells Proportion T-cells Proportion T-cells CD8+ Proportion
p1404_1.CEL          3.781402    6.999977      6.360039         0.06186246         0.06875119              0.03787857
2.CEL                4.208606    7.738854      6.389830         0.05533522         0.06881479              0.03339669
3.CEL                4.241815    7.489493      7.581455         0.07189161         0.08189133              0.04527415
4.CEL                3.923774    6.361952      6.167985         0.06264227         0.06954117              0.04298331
5.CEL                3.857191    6.142944      5.650759         0.06077186         0.07268853              0.04427215
6.CEL                4.329897    6.518604      5.913825         0.06137798         0.07304742              0.03825367
            NK-cells Proportion Cytotoxicity Score Proportion Neutrophils Proportion Monocytic lineage Proportion
p1404_1.CEL          0.06091409                    0.06799869             0.06560065                   0.08246773
2.CEL                0.05547191                    0.06685569             0.06530071                   0.09697717
3.CEL                0.05596475                    0.06721508             0.06095876                   0.08122769
4.CEL                0.06332778                    0.07044076             0.06771552                   0.08492636
5.CEL                0.06416664                    0.07127466             0.07119628                   0.08321033
6.CEL                0.06141811                    0.07056355             0.06763785                   0.08244950
            Macrophages Proportion M2 macrophage Proportion Myeloid Dendritic Cells Proportion Endothelial cells Proportion
p1404_1.CEL             0.09330604               0.07331228                         0.05774229                   0.07283472
2.CEL                   0.11238690               0.09535104                         0.05540084                   0.06763891
3.CEL                   0.08921869               0.07042817                         0.06317627                   0.06869252
4.CEL                   0.09743794               0.07307324                         0.05831487                   0.07383063
5.CEL                   0.09614560               0.06700058                         0.06408724                   0.07521367
6.CEL                   0.09475976               0.07030559                         0.05817439                   0.08317936
            Fibroblasts Proportion Smooth muscle Proportion
p1404_1.CEL              0.1348287                0.1225026
2.CEL                    0.1243755                0.1026946
3.CEL                    0.1212859                0.1227751
4.CEL                    0.1197079                0.1160582
5.CEL                    0.1197849                0.1101875
6.CEL                    0.1252254                0.1136074
```

## Plotting
I've also tried the implemented plotting function and it returns an error related to incorrect  number of dimension, this is the output:

```
> results <- predict_LundTax2023(Lund2017, 
+                                include_data = TRUE, 
+                                impute = TRUE, 
+                                verbose = TRUE)

[1] "Calculating Proliferation"
Proliferation score: 6/138 genes are missing from the data.
[1] "NUSAP1"   "PRC1"     "FAM72D"   "FAM72A"   "RNASEH2A" "POLA2"   
[1] "Calculating G3 score"
Loading required package: multiclassPairs
These genes are not found in the data:
ENSG00000109654 ENSG00000290292 ENSG00000115561 ENSG00000182768 ENSG00000139372 ENSG00000102172 ENSG00000292348 ENSG00000188610 ENSG00000125885 ENSG00000196550 ENSG00000013364 ENSG00000120963 ENSG00000173207 ENSG00000137804 ENSG00000133794 ENSG00000280789
Gene names should as rownames and sample names as columns!
Check the genes in classifier object to see all the needed genes.
Check if '-' or ',' symbols in the gene names in your data. You may need to change it to '_' or '.'
Missed genes will be imputed to the closest class for each sample!
These genes have NAs:
ENSG00000109654 ENSG00000290292 ENSG00000115561 ENSG00000182768 ENSG00000139372 ENSG00000102172 ENSG00000292348 ENSG00000188610 ENSG00000125885 ENSG00000196550 ENSG00000013364 ENSG00000120963 ENSG00000173207 ENSG00000137804 ENSG00000133794 ENSG00000280789
These genes will be imputed to the closest class for each sample with NAs
[1] "Calculating HG score"
Loading required package: multiclassPairs
These genes are not found in the data:
ENSG00000224531 ENSG00000104889 ENSG00000108106 ENSG00000172236 ENSG00000147155 ENSG00000198300 ENSG00000125885 ENSG00000149300 ENSG00000284479 ENSG00000069764 ENSG00000120669 ENSG00000171489 ENSG00000113522 ENSG00000228727 ENSG00000187838 ENSG00000269955
Gene names should as rownames and sample names as columns!
Check the genes in classifier object to see all the needed genes.
Check if '-' or ',' symbols in the gene names in your data. You may need to change it to '_' or '.'
Missed genes will be imputed to the closest class for each sample!
These genes have NAs:
ENSG00000224531 ENSG00000104889 ENSG00000108106 ENSG00000172236 ENSG00000147155 ENSG00000198300 ENSG00000125885 ENSG00000149300 ENSG00000284479 ENSG00000069764 ENSG00000120669 ENSG00000171489 ENSG00000113522 ENSG00000228727 ENSG00000187838 ENSG00000269955
These genes will be imputed to the closest class for each sample with NAs
[1] "Calculating Progression score"
Progression score: 28/200 genes are missing from the data.
 [1] "CCDC150"       "TATDN2"        "FAM72A"        "FAM72B"        "ZNF706"        "ESPL1"         "LOXL2"         "PCDHB3"       
 [9] "ZNF670-ZNF695" "UBE2D3"        "TSPYL1"        "EEF1B2"        "LY75"          "RPL26"         "EEF1A1"        "ZNF816"       
[17] "TMBIM4"        "RPL27A"        "PFDN5"         "RPS27"         "RPS3A"         "EIF4B"         "TPT1"          "SLC35A1"      
[25] "RPL13A"        "LDHB"          "EEF1G"         "BORCS7"       
[1] "Calculating Prostate score"
Prostate score: 3/17 genes are missing from the data.
[1] "POTEI"  "NKX3-1" "PDE11A"
[1] "Calculating Immune scores"
Immune scores: 3/71 genes are missing from the data.
[1] "KIR2DL3" "KLRK1"   "FCGR3B" 
[1] "Merging scores"

> plot_signatures(results)
Error in signatures[which(signatures$Signature == "early_cell_cycle."),  : 
  incorrect number of dimensions
```

Lastly, could you update my role permissions in this group so I can access the settings of this repo? You can find out how to do this [here](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-teams-and-people-with-access-to-your-repository#changing-permissions-for-a-team-or-person)

mattssca commented:
I was curious how the function `predict_LundTax2023` filters genes based on hgnc_symbols. I wanted to look into this because it seemed that in many cases missing genes were reported (with `verbose = TRUE`). I looked at the code related to when gene symbol is set to `hgcn_symbol` in this function. I might be missing something here but isn't the goal in this chunk to subset the genes in the incoming matrix to genes available in the bundled gene table? If so, there might be room for improvement. I did some tests trying to understand this better. 
My tests:

```
> #old way
> #set up test environment
> load("~/LU/repos/LundTaxonomy2023Classifier-elena.dev/data/Lund2017.rda")
> load("~/LU/repos/LundTaxonomy2023Classifier-elena.dev/data/gene_info_lund.rda")
> 
> #subset row names to column
> rownames(gene_info_lund) <- gene_info_lund[["hgnc_symbol"]]
> 
> #get all the genes from the incoming dataset that are also described in the gene list
> int_genes <- rownames(Lund2017)[which(rownames(Lund2017) %in% gene_info_lund[["hgnc_symbol"]])]
> 
> #set row names to hgnc_symbol and subset to genes in the bundled list with genes
> rownames(Lund2017)[which(rownames(Lund2017) %in% gene_info_lund[["hgnc_symbol"]])] <- gene_info_lund[int_genes,"hgnc_symbol"]
> 
> #checks
> #how many rows
> nrow(Lund2017)
[1] 15697
> 
> #subset row names to column (for checking purposes)
> Lund2017 <- tibble::rownames_to_column(Lund2017, "hgnc_symbol")
> 
> #are all genes in the filtered dataset present in the gene_info_lund dataset (should be TRUE)
> all(Lund2017$hgnc_symbol %in% gene_info_lund$hgnc_symbol)
[1] FALSE
> 
> #are all genes in the gene_info_lund dataset present in the filtered dataset (should be FALSE)
> all(gene_info_lund$hgnc_symbol %in% Lund2017$hgnc_symbol)
[1] FALSE

> #new way
> #set up test environment
> load("~/LU/repos/LundTaxonomy2023Classifier-elena.dev/data/Lund2017.rda")
> load("~/LU/repos/LundTaxonomy2023Classifier-elena.dev/data/gene_info_lund.rda")
> 
> #subset row names to column
> Lund2017 <- tibble::rownames_to_column(Lund2017, "hgnc_symbol")
> 
> #filter out genes that are not in gene_info_lund
> filtered_lund_new = dplyr::filter(Lund2017, hgnc_symbol %in% gene_info_lund$hgnc_symbol)
> 
> #set row names to hgnc_symbol
> row.names(filtered_lund_new) <- filtered_lund_new$hgnc_symbol
> 
> #remove hgnc_symbol column
> filtered_lund_new[1] <- NULL
> 
> #checks
> #how many rows
> nrow(filtered_lund_new)
[1] 1816
> 
> #are all genes in the filtered dataset present in the gene_info_lund dataset (should be TRUE)
> all(filtered_lund_new$hgnc_symbol %in% gene_info_lund$hgnc_symbol)
[1] TRUE
> 
> #are all genes in the gene_info_lund dataset present in the filtered dataset (should be FALSE)
> all(gene_info_lund$hgnc_symbol %in% filtered_lund_new$hgnc_symbol)
[1] FALSE
```
My proposed solution ensures only genes in the annotated list are kept in the incoming expression matrix. Is this the overall goal?

elenaara commented:
Hi Adam, 
For the hgnc_symbol part my goal is just to switch between hgnc symbol and ensembl ID when necessary. You can of course also keep only the genes present in the matrix and we can add that for simplicity instead of having the whole matrix which is not necessary.

The symbols are a problem here because different parts of the code use different types of gene symbols. When we started working on this I used the data as we had it with gene symbols and developed the subtype classifier, but after that I realized having everything based on ensembl IDs was much more reliable and convenient to manage and translate between different gene symbols, since 1 ensembl corresponds to 1 hgnc but 1 hgnc can have different ensembls or be updated/changed more frequently.
The newest classifiers and signatures are built using ensembl IDs, since I hoped everything could in the future switch to ensembl, maybe in future updates of the classifier. But right now in different steps gene names need to be translated depending on what the classifier needs (hgnc for subtype classifier and ensembl for the grade classifier).

mattssca commented:
Awesome, thanks for clarifying Elena!

elenaara committed:
Update functions and grade classifiers

elenaara commented:
Thank you for your comments!
I uploaded a new version.

- Deleted `visualize_scores.R`.
- Fixed the `verbose` error.
- Replaced all of the `updated_signatures` for `signatures`. The new `signatures` is the updated version
- First column in the scores dataframe is now `ProliferationScore`.
- Hardcoded paths removed. This was mainly related to the grade classifiers since I was not working on this function. Now it is updated and I also updated the names of the classifier objects so they are clearer.
- Impute error in test code: I don't think this is related to the gene list. The example code should include `impute = TRUE`.
- Plotting error: I updated the part where `signatures` is used and now it should work fine (it does for me).


## Potentially inaccurate values in the scores data frame.

Here the `PossibleProstate` values are based on a threshold that can be set with `threshold_prostate = 3`. The default is 3, decided based on the Uroscanseq data, but could be set a bit higher. Here you can set `threshold_prostate = 5` since all values are around 4, but also some genes are missing so I am not sure how reliable it would be. I added this with the intention that anyone could adjust the threshold to their dataset, but the `PossibleProstate` column could also be removed and only have the numerical score. 
This method is also used in the ` ProgressionRisk` column, assigned HR or LR based on a threhold applied to the `ProgressionScore`.

The error in the `Immune141UP`and `Stromal141UP` scores was caused by a wrong column name, and is related to the gene IDs. Since all the other signatures in this function have 2 possible gene IDs, their columns are named `hgnc_symbol` or `ensembl_gene_id`. The 141UP signatures only have hgnc so for now I made everything use the hgnc_symbol version. I think I tried to match the hgnc symbols in this signature to ensembl and it was not very straightforward, but I will try again, or we can change the code if you think of other solutions. 







> Lastly, could you update my role permissions in this group so I can access the settings of this repo? You can find out how to do this [here](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-teams-and-people-with-access-to-your-repository#changing-permissions-for-a-team-or-person)

I cannot change any roles or permissions in the settings, I just see everyone as "collaborator" and I can only add or remove collaborators. I think it is because it is not an organization account, maybe you guys there can change that setting and decide who is the owner of the organization then.

mattssca commented:
Amazing Elena, I will clone this branch and give the package another go.

mattssca commented on R/predict_functions.R:
```diff
@@ -1,806 +1,804 @@
+  classifier_GRADE3 <- LundTax2023Classifier::classifier_GRADE3
+  
+  # WHO 2004/2016 (HG vs LG)
+  load("C:/Users/earam/LBCG/LundTaxonomy2023Classifier/data/classifier_HG.rda")
```
Note, another local path

elenaara replied:
Removed

mattssca commented on R/predict_functions.R:
```diff
@@ -1,806 +1,804 @@
+  # Proliferation #
+  # Debug
+  if (verbose) print("Calculating Proliferation")
+  results_proliferation <- ratio_score(Data = D,
```
I can see that in the main function `lund_scores` the `logTransform = TRUE` but in all the internal calls of these scoring functions the same parameter is hardcoded to FALSE. Wouldn't this mean that if the user sets this parameter to TRUE in the main function, all helper function would still be run with this as FALSE?

I think it is better to set the internal calls inside this function to `logTransform = logTransform` to still give the user 100% of this parameter.

elenaara replied:
I just did it like this so that it does not log transform for each of the calls, it does it at the beggining of the function and then it is not needed for each of the scores since the data is already log transformed. Maybe there is a better way to do it but I wouldn't set it like this since it will log for every score. For now I removed the `logTransform = FALSE` from the calls since `FALSE` is already the default.

mattssca commented on R/predict_functions.R:
```diff
@@ -1,806 +1,804 @@
+                        adjust = FALSE,
+                        adj_factor = 5.1431,
+                        verbose = FALSE,
+                        ... # arguments to pass to the ranger functions (add impute = TRUE here if gene are missing)
```
"ranger function", do you mean the `predict_RF` function?

elenaara replied:
I took this from the multiclassPairs function, since predict_RF calls ranger inside and there is were the extra arguments will be used.

mattssca commented:
Hi, I finished my tests and everything looks good, besides the few minor comments I had above. If you have time, you could address these items and after this I can approve the PR and merge it with master. Thanks!

elenaara committed:
Update predict functions

elenaara commented:
I made some changes, but should this be merged already? I thought we should have everything finished to have in the main branch. I also thought it this would go together with the new paper in preparation, I am not sure how finished it is.

mattssca commented:
Thanks for looking into this Elena. I will do some final tests of the updates code today. Sure, we can hold off with merging until everything is in place. Cheers!
